cpp/src/arrow/dataset/dataset_test.cc

```diff
@@ -58,5 +59,54 @@ TEST_F(TestSimpleDataSource, GetFragments) {
   AssertDataSourceEquals(reader.get(), &source);
 }
 
+TEST(TestProjector, AugmentWithNull) {
+  constexpr int64_t kBatchSize = 1024;
+
+  auto from_schema = schema({field("f64", float64())});
+
+  auto batch = GetRecordBatch(kBatchSize, from_schema);
+
+  auto to_schema = schema({field("i32", int32()), field("f64", float64())});
+
+  RecordBatchProjector projector(default_memory_pool(), to_schema);
+
+  std::shared_ptr<Array> null_i32;
+  ASSERT_OK(MakeArrayOfNull(int32(), batch->num_rows(), &null_i32));
+  auto expected_batch =
+      RecordBatch::Make(to_schema, batch->num_rows(), {null_i32, batch->column(0)});
+
+  std::shared_ptr<RecordBatch> reconciled_batch;
+  ASSERT_OK(projector.Project(*batch, &reconciled_batch));
+
+  AssertBatchesEqual(*expected_batch, *reconciled_batch);
+}
+
+TEST(TestProjector, AugmentWithScalar) {
+  constexpr int64_t kBatchSize = 1024;
+  constexpr int64_t kScalarValue = 3;
+
+  auto from_schema = schema({field("f64", float64())});
```

Comment:
I'd use `schema({field("f64", float64()), field("b", bool())});` so you also test re-ordering and dropping a column.